Security group: Fix Display

gildub opened the pull request:
The security group rules are not displayed properly any more, this seems to be related with  "selectpicker-for-select-tag" => "", removing them restores normal behaviour of ng-repeat and ng-options

Also
* Removes `ng-selected` which is not needed and not recommended along `ng-repeat` which takes care of it.
* Changes `security_groups_list` retrieval to proper synchronous flow. 
* Simplifies copy of data to Model using Object.assign and relationships in form.
* Better reset using `cloneDeep` and cleanup for firewall_rules.

https://bugzilla.redhat.com/show_bug.cgi?id=1507915

gildub commented:
@dclarizio, @martinpovolny, @himdel, this is high priority because Security Groups Rules CRUD is a new feature to the new release. This was working before some change impacted it (see description). Thanks

gildub commented:
@miq-bot add_label bug

himdel commented:
@gildub `selectpicker-for-select-tag` is the directive used to initialize the selectpicker component. It is defined in `app/assets/javascripts/directives/selectpickerForSelectTag.js` .. and either `selectpicker-for-select-tag` or `pf-select` (from angular-patternfly 3.23) are needed, otherwise the selects are styled wrong.

(Also, you may need to rebase, looks like Travis is failing more than it should :).)

gildub commented:
@himdel, yeah I saw Travis issues last week, BTW thanks for tackling it. Rebased. Thanks for the details about `selectpicker`. My tests shows that when using it it breaks the table, I'll try against `pf-select` to see if that's any better.

himdel commented:
> My tests shows that when using it it breaks the table

Do you have a screenshot please? (Or, is the breakage visual or..?)

gildub commented:
@himdel, the symptoms are that neither the data model are showing up and the select options are not available in the drop downs. 

This is the actual broken situation:

![fw-rules-broken](https://user-images.githubusercontent.com/1901741/32580949-8de9de5e-c533-11e7-87a7-db54fe87b72a.png)

And this is what it should look like with current patch, which actually fix other things as well but the `selectpicker-for-select-tag` removal fixes the display issue:

![fw-rules-normal](https://user-images.githubusercontent.com/1901741/32582352-11626110-c53a-11e7-86c1-997f3c7b0bfc.png)

ZitaNemeckova commented:
BZ has `blocker?` flag so adding corresponding label
@miq-bot add_label blocker

himdel commented on app/assets/javascripts/controllers/security_group/security_group_form_controller.js:
```diff
@@ -20,33 +18,27 @@ ManageIQ.angular.app.controller('securityGroupFormController', ['securityGroupFo
+        miqService.sparkleOn();
+        API.get("/api/security_groups/" + securityGroupFormId + "?attributes=name,ext_management_system.name,description,cloud_tenant.name,firewall_rules").then(function(data) {
+          Object.assign(vm.securityGroupModel, data);
+          vm.securityGroupModel.firewall_rules_delete = false;        vm.securityGroupModel.name = data.name;
```
why that `vm.securityGroupModel.name = data.name;` ? Isn't that handled by the assign too? (But if not, formatting..)

gildub replied:
Yes, that's a residual cut mistake, removed, thanks!

himdel commented:
OK, tested in the UI, this is definitely better than before :).

I think loading the security group data should not have to wait for the security group list query, but if this fixes a bug.. :). Although.. maybe this is no longer needed since 

Also, the table is too wide now, if you have security groups with long names - previously, the select was limited to a specific width, now it grows so wide I can't see the right side.

---

I do agree that you don't want the old approach with `selectpicker-for-select-tag`.

But, I don't see why you can't use that `pf-select`...

From what I can tell, it actually makes the table a bit better:

current state (plain `<select>`):

![select](https://user-images.githubusercontent.com/289743/32947235-d0c0ce4e-cb92-11e7-9b35-4712725fca55.png)

`pf-select`:

![pf-select](https://user-images.githubusercontent.com/289743/32947245-dd5d5adc-cb92-11e7-9164-855d9cb81dca.png)

himdel commented:
Ok, some progres..

removing that `"data-width"                  => "auto",` from those select elements makes `pf-select` do the right thing:

![fixed](https://user-images.githubusercontent.com/289743/32947779-a2fa06ae-cb94-11e7-8b5a-40078126cee5.png)

himdel commented on app/assets/javascripts/controllers/security_group/security_group_form_controller.js:
```diff
@@ -20,33 +18,27 @@ ManageIQ.angular.app.controller('securityGroupFormController', ['securityGroupFo
-      miqService.sparkleOn();
-      API.get("/api/security_groups/" + securityGroupFormId + "?attributes=name,ext_management_system.name,description,cloud_tenant.name,firewall_rules").then(function(data) {
-        vm.securityGroupModel.name = data.name;
-        vm.securityGroupModel.ems_name = data.ext_management_system.name;
```
It looks like you're removing `ems_name` completely - is that intentional?

The same goes for `cloud_tenant_name`.

gildub replied:
The name is only for the display.

himdel commented:
Ok, this is the diff for the style part..

```diff
diff --git a/app/views/security_group/_fw_rules.haml b/app/views/security_group/_fw_rules.haml
index 51bca0bd1..e9667cc93 100644
--- a/app/views/security_group/_fw_rules.haml
+++ b/app/views/security_group/_fw_rules.haml
@@ -23,32 +23,32 @@
       %tr{"ng-repeat" => "fw_rule in vm.securityGroupModel.firewall_rules",
           "ng-form"   => "rowForm",
           "ng-if"     => "!fw_rule.deleted"}
-        %td.table-view-pf-select
+        %td
           %select{"name"                        => "direction",
-            "data-width"                  => "auto",
-            "ng-model"                    => "vm.securityGroupModel.firewall_rules[$index].direction",
-            "ng-options"                  => "direction for direction in vm.directions",
-            "required"                    => ""}
-        %td.table-view-pf-select
+                  "ng-model"                    => "vm.securityGroupModel.firewall_rules[$index].direction",
+                  "ng-options"                  => "direction for direction in vm.directions",
+                  "required"                    => "",
+                  'pf-select'                   => true}
+        %td
           %select{"name"                        => "network_protocol",
-            "data-width"                  => "auto",
-            "ng-model"                    => "vm.securityGroupModel.firewall_rules[$index].network_protocol",
-            "ng-options"                  => "network_protocol for network_protocol in vm.networkProtocols"}
+                  "ng-model"                    => "vm.securityGroupModel.firewall_rules[$index].network_protocol",
+                  "ng-options"                  => "network_protocol for network_protocol in vm.networkProtocols",
+                  'pf-select'                   => true}
             %option{"value" => ""}
               = "<#{_('Choose')}>"
-        %td.table-view-pf-select
+        %td
           %select{"name"                        => "host_protocol",
-            "data-width"                  => "auto",
-            "ng-model"                    => "vm.securityGroupModel.firewall_rules[$index].host_protocol",
-            "ng-options"                  => "host_protocol for host_protocol in vm.hostProtocols"}
+                  "ng-model"                    => "vm.securityGroupModel.firewall_rules[$index].host_protocol",
+                  "ng-options"                  => "host_protocol for host_protocol in vm.hostProtocols",
+                  'pf-select'                   => true}
             %option{"value" => ""}
               = "<#{_('Choose')}>"
-        %td.table-view-pf-select
+        %td
           %select{"name"                        => "source_security_group_id",
-            "data-width"                  => "auto",
-            "ng-model"                    => "vm.securityGroupModel.firewall_rules[$index].source_security_group_id",
-            "ng-options"                  => "sg.id as sg.name + ' - ' + sg.ems_ref for sg in vm.security_groups_list",
-            "ng-change"                   => "vm.securityGroupModel.firewall_rules[$index].source_ip_range = null"}
+                  "ng-model"                    => "vm.securityGroupModel.firewall_rules[$index].source_security_group_id",
+                  "ng-options"                  => "sg.id as sg.name + ' - ' + sg.ems_ref for sg in vm.security_groups_list",
+                  "ng-change"                   => "vm.securityGroupModel.firewall_rules[$index].source_ip_range = null",
+                  'pf-select'                   => true}
             %option{"value" => ""}
               = "<#{_('Choose')}>"
         %td
```

Does that work for you? :)

(It also removes the `table-view-pf-select` class, since that's intended for the checkbox column in lists.)

himdel commented:
> I think loading the security group data should not have to wait for the security group list query, but if this fixes a bug.. :).

OK, from the BZ, I can't tell what the bug really was, but if having the requests go one by one is not part of the fix, I'd rather you made them run in parallel..

Does this work for you..?

```diff
diff --git a/app/assets/javascripts/controllers/security_group/security_group_form_controller.js b/app/assets/javascripts/controllers/security_group/security_group_form_controller.js
index c0b498731..290b6665d 100644
--- a/app/assets/javascripts/controllers/security_group/security_group_form_controller.js
+++ b/app/assets/javascripts/controllers/security_group/security_group_form_controller.js
@@ -1,4 +1,4 @@
-ManageIQ.angular.app.controller('securityGroupFormController', ['securityGroupFormId', 'miqService', 'API', function(securityGroupFormId, miqService, API) {
+ManageIQ.angular.app.controller('securityGroupFormController', ['securityGroupFormId', 'miqService', 'API', '$q', function(securityGroupFormId, miqService, API, $q) {
   var vm = this;
 
   var init = function() {
@@ -18,23 +18,38 @@ ManageIQ.angular.app.controller('securityGroupFormController', ['securityGroupFo
     vm.newRecord = securityGroupFormId === "new";
     vm.saveable = miqService.saveable;
 
-    API.get("/api/security_groups/?expand=resources&attributes=ems_ref,id,name").then(function(data) {
-      vm.security_groups_list = data.resources;
-    }).then(function() {
-      if (vm.newRecord) {
-        vm.afterGet = true;
-        vm.modelCopy = angular.copy( vm.securityGroupModel );
-      } else {
-        miqService.sparkleOn();
-        API.get("/api/security_groups/" + securityGroupFormId + "?attributes=name,ext_management_system.name,description,cloud_tenant.name,firewall_rules").then(function(data) {
+    if (vm.newRecord) {
+      vm.afterGet = true;
+      vm.modelCopy = angular.copy( vm.securityGroupModel );
+      getSecurityGroups()
+        .catch(miqService.handleFailure);
+    } else {
+      miqService.sparkleOn();
+
+      $q.all([getSecurityGroup(securityGroupFormId), getSecurityGroups()])
+        .then(function(array) {
+          var data = array[0];
+
           Object.assign(vm.securityGroupModel, data);
-          vm.securityGroupModel.firewall_rules_delete = false;        vm.securityGroupModel.name = data.name;
+          vm.securityGroupModel.firewall_rules_delete = false;
           vm.afterGet = true;
           vm.modelCopy = _.cloneDeep(vm.securityGroupModel);
+
           miqService.sparkleOff();
-        }).catch(miqService.handleFailure);
-      }
-    }).catch(miqService.handleFailure);
+        })
+        .catch(miqService.handleFailure);
+    }
+  };
+
+  var getSecurityGroup = function(id) {
+    return API.get("/api/security_groups/" + id + "?attributes=name,ext_management_system.name,description,cloud_tenant.name,firewall_rules");
+  };
+
+  var getSecurityGroups = function() {
+    return API.get("/api/security_groups/?expand=resources&attributes=ems_ref,id,name")
+      .then(function(data) {
+        vm.security_groups_list = data.resources;
+      });
   };
 
   vm.addClicked = function() {
```

himdel commented:
.. alternately + this, if the idea was to wait for securityGroups with afterGet, in the `new` case...

```diff
diff --git a/app/assets/javascripts/controllers/security_group/security_group_form_controller.js b/app/assets/javascripts/controllers/security_group/security_group_form_controller.js
index 290b6665d..96a9cc8f0 100644
--- a/app/assets/javascripts/controllers/security_group/security_group_form_controller.js
+++ b/app/assets/javascripts/controllers/security_group/security_group_form_controller.js
@@ -19,9 +19,11 @@ ManageIQ.angular.app.controller('securityGroupFormController', ['securityGroupFo
     vm.saveable = miqService.saveable;
 
     if (vm.newRecord) {
-      vm.afterGet = true;
-      vm.modelCopy = angular.copy( vm.securityGroupModel );
       getSecurityGroups()
+        .then(function() {
+          vm.afterGet = true;
+          vm.modelCopy = angular.copy( vm.securityGroupModel );
+        })
         .catch(miqService.handleFailure);
     } else {
       miqService.sparkleOn();
```

gildub commented:
@himdel, thanks for the investigation.

The initial display regression issue is addressed by the first commit:  https://github.com/ManageIQ/manageiq-ui-classic/pull/2543/commits/99aac08dde4a87af22fd9713c5efbb0740d3b5f5. Sorry I didn't make that obvious enough. 

I've restored the `pf-select` part and removed the auto-width, that works bit width size are not optimal.

I'm not sure why I shared the security-groups-list (getSecurityGroups) for both, it's only needed for edit, fixed that too.

himdel commented:
OK, ignoring the parallel request issue and merging.

simaishi commented:
Gaprindashvili backport details:

```
$ git log -1
commit 94cb7a03070cbdfb4dd7dded02f4a14f11126666
Author: Martin Hradil <himdel@seznam.cz>
Date:   Mon Nov 20 13:50:55 2017 +0000

    Merge pull request #2543 from gildub/security_group_fix
    
    Security group: Fix Display
    (cherry picked from commit 87eee0f227529e27cc87bcc6e566303a07c94a06)
```
